src/playlist-loader.js

```diff
@@ -379,57 +389,104 @@
       return 0;
     }
 
-    // 1) Walk backward until we find the latest segment with timeline
+    // 1) Walk backward until we find the first segment with timeline
     // information that is earlier than `time`
-    targetDuration = this.media_.targetDuration || 10;
-    i = this.media_.segments.length;
-    while (i--) {
+    for (i = lastSegment; i > 0; i--) {
       segment = this.media_.segments[i];
       if (segment.end !== undefined && segment.end <= time) {
-        time -= segment.end;
+        startIndex = i + 1;
+        knownStart = segment.end;
+        if (startIndex >= numSegments) {
+          // Somehow the last segment claims to end *before* the time we are
+          // searching for so just return it
+          return lastSegment;
+        }
         break;
       }
       if (segment.start !== undefined && segment.start < time) {
-
         if (segment.end !== undefined && segment.end > time) {
           // we've found the target segment exactly
           return i;
         }
+        startIndex = i;
+        knownStart = segment.start;
+        break;
+      }
+    }
 
-        time -= segment.start;
-        time -= segment.duration || targetDuration;
-        if (time < 0) {
-          // the segment with start information is also our best guess
-          // for the momment
+    // 2) Walk forward until we find the first segment with timeline
+    // information that is greater than `time`
+    for (i = 0; i < this.media_.segments.length; i++) {
```

Comment:
I think you should use `numSegments` as your end condition here for consistency or get rid of the alias entirely. I kind of lean towards dropping `numSegments` since we already have a ton of function variables in scope already but it's your call.